### PR TITLE
refactor(timelock): change submittedAt to validAt

### LINKED ISSUE #267: Unpredictable timelock mechanism
The issue highlighted in https://github.com/cantinasec/review-morpho-blue-1/pull/68#discussion_r1367673403 seems to be fixed with #259. But actually this is a deeper issue where timelock of pending values are variable which can be unexpected. Consider the following scenario:
1. Let's say the timelock is initially of 1 week, mark the current time as `t`
2. People agree to change the timelock to 1 day, a pending value is submitted at `t`, with the intent that it is effective at `t + 7`
3. 3 days later, at `t + 3`, people agree to increase the fee, with the intent that it is effective at `t + 10`
4. At `t + 7` the timelock is accepted
5. After that (starting at `t + 7` and beyond) the fee can be increased

I think that we should change the `submittedAt` field of pending values to a `validFrom` field that would store `block.timestamp + timelock` and the moment where the pending value is set

### PR DESCRIPTION
**Pull request by Rubilmax**
- Fixes #267

### CODE REVIEW

**Review by Jean-Grimal — APPROVED** [review]
I'd rather stay with the current version though

File: src/libraries/PendingLib.sol [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+struct PendingAddress {
+    /// @notice The pending value to set.
+    address value;
+    /// @notice The timestamp at which the value was submitted.
+    uint96 validAt;
+}
+
+/// @title ConstantsLib
```
Comment (MerlinEgalite) [inline src/libraries/PendingLib.sol:18]:
```suggestion
/// @title PendingLib
```

File: src/libraries/PendingLib.sol [hunk 2]
```diff
@@ -0,0 +1,38 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity ^0.8.0;
+
+struct PendingUint192 {
+    /// @notice The pending value to set.
+    uint192 value;
+    /// @notice The timestamp at which the value was submitted.
```
Comment (MerlinEgalite) [inline src/libraries/PendingLib.sol:7]:
```suggestion
    /// @notice The timestamp at which the pending value becomes valid.
```

File: src/libraries/PendingLib.sol [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    uint64 validAt;
+}
+
+struct PendingAddress {
+    /// @notice The pending value to set.
+    address value;
+    /// @notice The timestamp at which the value was submitted.
```
Comment (MerlinEgalite) [inline src/libraries/PendingLib.sol:14]:
```suggestion
    /// @notice The timestamp at which the pending value becomes valid.
```

**Review by MerlinEgalite — CHANGES_REQUESTED** [review]

**Comment by Rubilmax** [comment]
> I'd rather stay with the current version though

But there's a bug with it, see the associated issue
What specifically don't you like about the new version? We can modify things to find a compromise

**Comment by Jean-Grimal** [comment]
> > I'd rather stay with the current version though
> 
> But there's a bug with it, see the associated issue What specifically don't you like about the new version? We can modify things to find a compromise

Oh yes sorry I was mistaken. I understand the issue now

**Review by MerlinEgalite — APPROVED** [review]

**Review by Jean-Grimal — APPROVED** [review]

**Review by MerlinEgalite — APPROVED** [review]

**Comment by MerlinEgalite** [comment]
@Rubilmax can you update this PR please?

**Review by MerlinEgalite — APPROVED** [review]

**Review by MerlinEgalite — CHANGES_REQUESTED** [review]
I think the events haven't been updated though

File: src/MetaMorpho.sol [hunk 4]
```diff
@@ -201,8 +202,11 @@ contract MetaMorpho is ERC4626, ERC20Permit, Ownable2Step, Multicall, IMetaMorph
         if (newTimelock > timelock) {
             _setTimelock(newTimelock);
         } else {
-            // Safe "unchecked" cast because newTimelock <= MAX_TIMELOCK.
-            pendingTimelock = PendingUint192(uint192(newTimelock), uint64(block.timestamp));
+            pendingTimelock.update(
+                // Safe "unchecked" cast because newTimelock <= MAX_TIMELOCK.
+                uint192(newTimelock),
+                timelock
+            );
```
Comment (QGarchery) [inline src/MetaMorpho.sol:209]:
```suggestion
            // Safe "unchecked" cast because newTimelock <= MAX_TIMELOCK.
            pendingTimelock.update(uint192(newTimelock), timelock);
```
just to make it more readable (personal opinion)

**Review by MerlinEgalite — APPROVED** [review]

File: src/interfaces/IMetaMorpho.sol
Comment (QGarchery) [inline src/interfaces/IMetaMorpho.sol:1]: There will be many conflicts with #272, maybe base this PR on top ?

File: src/libraries/PendingLib.sol [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        // Safe "unchecked" cast because timelock <= MAX_TIMELOCK.
+        pending.validAt = uint64(block.timestamp) + uint64(timelock);
+    }
+
+    /// @dev Updates `pending`'s value to `newValue` and its corresponding `validAt` timestamp.
+    /// @dev Assumes `timelock` <= `MAX_TIMELOCK`.
+    function update(PendingAddress storage pending, address newValue, uint256 timelock) internal {
+        pending.value = newValue;
+        // Safe "unchecked" cast because timelock <= MAX_TIMELOCK.
+        pending.validAt = uint64(block.timestamp) + uint64(timelock);
```
Comment (QGarchery) [inline src/libraries/PendingLib.sol:36]:
```suggestion
        pending.validAt = uint96(block.timestamp) + uint96(timelock);
```
because otherwise there will be another implicit cast to `uint96` anyway

**Review by QGarchery — COMMENTED** [review]
Adding 6 tests :muscle:

A few minor things to change but I like this implementation !

**Comment by Rubilmax** [comment]
> Adding 6 tests 💪
> 
> A few minor suggestions but I like this implementation !

4 of them will actually be removed in #256

File: src/MetaMorpho.sol
Comment (QGarchery) [inline src/MetaMorpho.sol:1]: The target branch of this PR should be changed to `refactor/interfaces-3`

File: src/libraries/PendingLib.sol (hunk 5 above)
Comment (QGarchery) [reply src/libraries/PendingLib.sol:36]: This is not resolved right ?

**Review by QGarchery — CHANGES_REQUESTED** [review]

File: src/libraries/PendingLib.sol (hunk 5 above)
Comment (Rubilmax) [reply src/libraries/PendingLib.sol:36]: Yes it is, by simply storing `validAt` as `uint64` ^^

File: src/libraries/PendingLib.sol [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
```
Comment (MerlinEgalite) [inline src/libraries/PendingLib.sol:36]:
```suggestion
        pending.validAt = uint64(block.timestamp + timelock);
```

Isn't possible?

**Review by MerlinEgalite — APPROVED** [review]

File: src/libraries/PendingLib.sol [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        pending.value = newValue;
+        // Safe "unchecked" cast because timelock <= MAX_TIMELOCK.
+        pending.validAt = uint64(block.timestamp + timelock);
+    }
+
+    /// @dev Updates `pending`'s value to `newValue` and its corresponding `validAt` timestamp.
+    /// @dev Assumes `timelock` <= `MAX_TIMELOCK`.
+    function update(PendingAddress storage pending, address newValue, uint256 timelock) internal {
+        pending.value = newValue;
+        // Safe "unchecked" cast because timelock <= MAX_TIMELOCK.
```
Comment (StErMi) [inline src/libraries/PendingLib.sol:35]: You are referencing something that is not defined in this library.

Comment (Rubilmax) [reply src/libraries/PendingLib.sol:35]: I think there are other places like this. Should I:
```suggestion
        // Safe "unchecked" cast because timelock <= ConstantsLib.MAX_TIMELOCK.
```

@morpho-org/onchain-team ?

Comment (MerlinEgalite) [reply src/libraries/PendingLib.sol:35]: I don't think it's worth it tbh

**Comment by StErMi** [comment]
@MerlinEgalite just to recap the changes

- the owner now cannot directly influence the timelock of existing pending proposals (already scheduled) by calling `submitTimelock(newTimelock)` where `newTimelock > timelock`
- the owner can still influence every pending proposal by executing `submit*`. He can also pass the very same already proposed value. 
- the owner can anyway revoke the pending proposal and propose a new one (not in this PR, but there's another PR that makes the `owner` part of the guardian role) 

PS: is there a PR that does not allow setting the same values for the current active pending proposals?
TLDR the main difference is that the duration of the timelock is stored in the pending proposal when it's created and `submitTimelock(newTimelock)` cannot directly influence an already scheduled pending proposal even if the `timelock` is instantly changed or `pendingTimelock` is approved.

**Comment by MerlinEgalite** [comment]
> * the owner now cannot directly influence the timelock of existing pending proposals (already scheduled) by calling `submitTimelock(newTimelock)` where `newTimelock > timelock`
> * the owner can still influence every pending proposal by executing `submit*`. He can also pass the very same already proposed value.
> * the owner can anyway revoke the pending proposal and propose a new one (not in this PR, but there's another PR that makes the `owner` part of the guardian role)

This is exact in the current state of this PR.

> PS: is there a PR that does not allow setting the same values for the current active pending proposals? TLDR the main difference is that the duration of the timelock is stored in the pending proposal when it's created and `submitTimelock(newTimelock)` cannot directly influence an already scheduled pending proposal even if the `timelock` is instantly changed or `pendingTimelock` is approved.

yes there's such PR here: https://github.com/morpho-org/metamorpho/pull/295
